kernel/sched.c

```diff
@@ -181,6 +181,55 @@ static ALWAYS_INLINE struct k_thread *_priq_dumb_mask_best(sys_dlist_t *pq)
 }
 #endif
 
+/* _current is never in the run queue until context switch on
+ * SMP configurations, see z_requeue_current()
+ */
+static inline bool should_queue_thread(struct k_thread *th)
+{
+	return !IS_ENABLED(CONFIG_SMP) || th != _current;
```

Comment:
>  **except** for the case of _current, which remains outside the global run queue until the point of a context switch.

A-hah.  I misread the condition.

    !IS_ENABLED(CONFIG_SMP) || th != _current;

does not mean "don't queue if SMP and *not* current thread.  It's equivalent to the **much** clearer expression:

    !(IS_ENABLED(CONFIG_SMP) && (th == _current));

which means "don't queue if SMP and *is* current thread".

Then it makes sense.

I wouldn't object to using the alternative expression here.